### PR TITLE
Allow curly around `@type` jsdoc to be optional

### LINKED ISSUE #19987: JSDoc Allow `{}` around typecasts/declarations to be optional
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Code**

```ts
        this._cachedTypeClasses[typeName] = /** @type function(new:Object) */ (object);
```

**Expected behavior:**
`object` cast to `function(new:Object)`

**Actual behavior:**
```
front_end/Runtime.js(398,24): error TS1138: Parameter declaration expected.
```
(Parse error on `function` keyword)

We've seen this before (and crashed on it), but choose not to make it not an error because it seemed wrong; however a lot of code that validates under closure compiler (namely the chrome devtools) appears to be written like this.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #19987

### CODE REVIEW

**Comment by weswigham** [comment]
For reference, this reverses the decision made in #16947 for this to be an error, given that we have a large sample which uses `@type` tags without curly braces that validates under closure compiler, I'd say allowing the braces to be omitted is more correct.

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -6131,6 +6131,14 @@ namespace ts {
                 return jsDocTypeExpression ? { jsDocTypeExpression, diagnostics } : undefined;
             }
 
+            function parseUnbracketedJSDocTypeExpression(): JSDocTypeExpression {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: This could be done in `parseJSDocTypeExpression` like we do with `requireBraces`, which happens to be no longer used!

**Comment by andy-ms** [comment]
Sorry for the ambiguous issue title -- the problem fixed in #16947 was that we would parse `number` as a comment and not a type.

**Comment by weswigham** [comment]
@andy-ms I've merged the implementations and removed the unused overload.

@sandersn Did you want to look over this before its merged? I believe you had comments about the jsdoc scanner changes.

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -6132,16 +6132,16 @@ namespace ts {
...
                 const result = <JSDocTypeExpression>createNode(SyntaxKind.JSDocTypeExpression, scanner.getTokenPos());
 
-                if (!parseExpected(SyntaxKind.OpenBraceToken) && requireBraces) {
-                    return undefined;
+                if (!braceless) {
+                    parseExpected(SyntaxKind.OpenBraceToken);
                 }
                 result.type = doInsideOfContext(NodeFlags.JSDoc, parseType);
-                parseExpected(SyntaxKind.CloseBraceToken);
+                if (!braceless) {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: I have a recommended change at https://github.com/Microsoft/TypeScript/commit/ff94f2da30da1c9567a8f01ee2a11f0e7352ef4f

Comment (weswigham) [reply src/compiler/parser.ts]: Done

**Pull request merged**